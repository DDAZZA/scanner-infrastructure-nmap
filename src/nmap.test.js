const { transform } = require('./nmap');

describe('nmap', () => {
    describe('transform', () => {
        beforeAll(() => {
            jest.mock('uuid/v4');
        });

        afterAll(() => {
            jest.unmock('uuid/v4');
        });

        it('should transform a empty host array into an empty port array', () => {
            const findings = transform([]);

            expect(findings).toEqual([]);
        });

        it('should transform a null host array into an empty port array', () => {
            const findings = transform(null);

            expect(findings).toEqual([]);
        });

        it('should transform results of a single host into a port array', () => {
            const [finding, ...otherFindings] = transform([
                {
                    hostname: 'securebox',
                    ip: '192.168.99.100',
                    mac: null,
                    openPorts: [
                        {
                            port: 22,
                            protocol: 'tcp',
                            service: 'ssh',
                            method: 'table',
                        },
                    ],
                    osNmap: null,
                },
            ]);

            expect(otherFindings).toEqual([]);
            expect(finding).toEqual({
                id: '49bf7fd3-8512-4d73-a28f-608e493cd726',
                name: 'ssh',
                description: 'Port 22 is open using tcp protocol.',
                category: 'Open Port',
                osiLayer: 'NETWORK',
                severity: 'INFORMATIONAL',
                reference: null,
                hint: null,
                location: 'tcp://192.168.99.100:22',
                attributes: {
                    port: 22,
                    ipAddress: '192.168.99.100',
                    protocol: 'tcp',
                    service: 'ssh',
                    method: 'table',
                    hostname: 'securebox',
                    macAddress: null,
                    operatingSystem: null,
                },
            });
        });

        xit('should transform results if a host has multiple open ports', () => {
            const findings = transform([
                {
                    hostname: 'securebox',
                    ip: '192.168.99.100',
                    mac: null,
                    openPorts: [
                        {
                            port: 22,
                            protocol: 'tcp',
                            service: 'ssh',
                            method: 'table',
                        },
                        {
                            port: 80,
                            protocol: 'tcp',
                            service: 'http',
                            method: 'table',
                        },
                    ],
                    osNmap: null,
                },
            ]);

            expect(findings).toEqual([
                {
                    port: 22,
                    protocol: 'tcp',
                    service: 'ssh',
                    method: 'table',
                    hostname: 'securebox',
                    ip: '192.168.99.100',
                    mac: null,
                    osNmap: null,
                },
                {
                    port: 80,
                    protocol: 'tcp',
                    service: 'http',
                    method: 'table',
                    hostname: 'securebox',
                    ip: '192.168.99.100',
                    mac: null,
                    osNmap: null,
                },
            ]);
        });

        xit('should transform results of multiple hosts into a port array', () => {
            const findings = transform([
                {
                    hostname: 'securebox',
                    ip: '192.168.99.100',
                    mac: null,
                    openPorts: [
                        {
                            port: 22,
                            protocol: 'tcp',
                            service: 'ssh',
                            method: 'table',
                        },
                    ],
                    osNmap: null,
                },
                {
                    hostname: 'test',
                    ip: '192.168.99.101',
                    mac: null,
                    openPorts: [
                        {
                            port: 80,
                            protocol: 'tcp',
                            service: 'http',
                            method: 'table',
                        },
                    ],
                    osNmap: null,
                },
            ]);

            expect(findings).toEqual([
                {
                    port: 22,
                    protocol: 'tcp',
                    service: 'ssh',
                    method: 'table',
                    hostname: 'securebox',
                    ip: '192.168.99.100',
                    mac: null,
                    osNmap: null,
                },
                {
                    port: 80,
                    protocol: 'tcp',
                    service: 'http',
                    method: 'table',
                    hostname: 'test',
                    ip: '192.168.99.101',
                    mac: null,
                    osNmap: null,
                },
            ]);
        });

        xit('should still kind of work if the openPorts attribute of the host is not an array', () => {
            const findings = transform([
                {
                    hostname: 'securebox',
                    ip: '192.168.99.100',
                    mac: null,
                    openPorts: null,
                    osNmap: null,
                },
            ]);

            expect(findings).toEqual([]);
        });
    });
});
