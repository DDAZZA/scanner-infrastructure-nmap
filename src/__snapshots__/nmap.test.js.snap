// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nmap worker should run multiple scans when multiple targets are configured 1`] = `
Object {
  "raw": Array [
    "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE nmaprun></nmaprun>",
    "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE nmaprun></nmaprun>",
  ],
  "result": Array [
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 631,
        "protocol": "tcp",
        "scripts": null,
        "service": "ipp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 631 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:631",
      "name": "ipp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 7778,
        "protocol": "tcp",
        "scripts": null,
        "service": "interwise",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 7778 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:7778",
      "name": "interwise",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 8080,
        "protocol": "tcp",
        "scripts": null,
        "service": "http-proxy",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 8080 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:8080",
      "name": "http-proxy",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 9200,
        "protocol": "tcp",
        "scripts": null,
        "service": "wap-wsp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 9200 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:9200",
      "name": "wap-wsp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 631,
        "protocol": "tcp",
        "scripts": null,
        "service": "ipp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 631 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:631",
      "name": "ipp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 7778,
        "protocol": "tcp",
        "scripts": null,
        "service": "interwise",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 7778 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:7778",
      "name": "interwise",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 8080,
        "protocol": "tcp",
        "scripts": null,
        "service": "http-proxy",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 8080 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:8080",
      "name": "http-proxy",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 9200,
        "protocol": "tcp",
        "scripts": null,
        "service": "wap-wsp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 9200 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:9200",
      "name": "wap-wsp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
  ],
}
`;

exports[`nmap worker should take parameters 1`] = `
Object {
  "raw": Array [
    "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE nmaprun></nmaprun>",
  ],
  "result": Array [
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 631,
        "protocol": "tcp",
        "scripts": null,
        "service": "ipp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 631 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:631",
      "name": "ipp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 7778,
        "protocol": "tcp",
        "scripts": null,
        "service": "interwise",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 7778 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:7778",
      "name": "interwise",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 8080,
        "protocol": "tcp",
        "scripts": null,
        "service": "http-proxy",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 8080 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:8080",
      "name": "http-proxy",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 9200,
        "protocol": "tcp",
        "scripts": null,
        "service": "wap-wsp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 9200 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:9200",
      "name": "wap-wsp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
  ],
}
`;

exports[`nmap worker should throw an error if a scan fails 1`] = `"Failed to execute nmap portscan."`;

exports[`nmap worker should work with a single target 1`] = `
Object {
  "raw": Array [
    "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE nmaprun></nmaprun>",
  ],
  "result": Array [
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 631,
        "protocol": "tcp",
        "scripts": null,
        "service": "ipp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 631 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:631",
      "name": "ipp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 7778,
        "protocol": "tcp",
        "scripts": null,
        "service": "interwise",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 7778 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:7778",
      "name": "interwise",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 8080,
        "protocol": "tcp",
        "scripts": null,
        "service": "http-proxy",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 8080 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:8080",
      "name": "http-proxy",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
    Object {
      "attributes": Object {
        "hostname": "localhost",
        "ip_address": "127.0.0.1",
        "mac_address": null,
        "method": "table",
        "operating_system": null,
        "port": 9200,
        "protocol": "tcp",
        "scripts": null,
        "service": "wap-wsp",
        "state": "open",
      },
      "category": "Open Port",
      "description": "Port 9200 is open using tcp protocol.",
      "id": "49bf7fd3-8512-4d73-a28f-608e493cd726",
      "location": "tcp://127.0.0.1:9200",
      "name": "wap-wsp",
      "osi_layer": "NETWORK",
      "severity": "INFORMATIONAL",
    },
  ],
}
`;
